### ISSUE TITLE
Potential starvation of Tier 2 writes with unthrottled clients

### ISSUE THREAD

**Issue by RaulGracia** [pravega/pravega#4158]
**Problem description**
Executing benchmarks in various Pravega deployments, we have found a common issue: Pravega performs much worse and exhibits higher instability when _clients write as fast as possible (unthrottled)_. Next, we describe a reproduction of this issue that can shed some light to understand the nature of the problem.

_Deployment_: The following experiment has been executed with Pravega 0.6 + RocksDB in memory (this removes one IO volume from the equation and provides better performance). The cluster deployed consists of 1 Segment Store and 3 Bookies (PKS, 4 large worker nodes, on-premise). The manifest used is the main one proposed internally, plus the new RocksDB flag to use in-memory option. 

_Benchmark_: We used a single pod to execute Pravega benchmark (v0.5). The benchmark executed for the first experiment is: `pravega-benchmark -controller tcp://xxx:9090 -stream perfeval1seg -segments 1 -producers 5 -time 600 -size 1000  -throughput 30`. The second experiment just removes the `-throughput` option, so clients are unthrottled.

_Observations_: In the following, we compare the behavior of the Pravega executing the benchmark with and without `throughput` option.
![anchorsteam_vs_frost_throttling_client](https://user-images.githubusercontent.com/717112/64462267-21df5380-d100-11e9-9fbf-1e636da546bd.jpg)

- When the client is throttled to 30MBps, the system is stable and the ingest rate of Pravega (`Segment Store Write Throughput`) and the writes to Tier 2 are balanced.
- With an unthrottled client, we observe that the ingest rate at the beginning of the experiment is higher than in the throttled case (35MBps-40MBps) for a short period of time. However, at the same time, the write rate to Tier 2 drops dramatically up to 10MBps.
- The imbalance between write throughput to Tier 1 and Tier 2 makes the cache to get filled progressively. Once it reaches the limit (3GB), an aggressive throttling activity comes into play, making the throughput perceived from the client low and very variable.
- Note that the latency of writes to Bookkeeper and Tier 2 are the same in both scenarios, so this effect cannot be attributed to any of these external components. Moreover, RocksDB in memory is faster and avoids any potential interference related to the cache storage volume. 

**Problem location**
Segment Store (storage writer, cache manager?).

**Suggestions for an improvement**
One possibility is that, when unthrottled clients write data to the Segment Store, Tier 1 components consume most of the resources (e.g., CPU) and make the Storage Writer to starve, thus reducing the throughput of writes to Tier 2 (i.e., throttling clients may give opportunities to the Storage Writer to get sufficient resources to keep balance). This hypothesis is supported by the above plot, as once the unthrottled writes to Tier 1 stop, there is a significant increase in the writes to Tier 2 that allow the cache to get emptied.

**Comment by gaddamas** [comment]
How about applying back pressure and congestion notification techniques which can help any kind of workloads achieve best possible throughput?

**Comment by andreipaduroiu** [comment]
Do you have metrics for this run?

**Comment by andreipaduroiu** [comment]
Do you have full metrics for this run?

From the graphs you're showing me, the cache write latency is pretty high (spikes to 10s). That metric (if you plotted the one with the same name) does not include throttling but just the amount of time it takes to write to the Read Index.

It would be helpful to see:
- cache put latency
- cache get latency
- The size of the Tier 1 log

**Comment by RaulGracia** [comment]
Related to #4164.

**Comment by RaulGracia** [comment]
Thanks @andreipaduroiu, yes these issues treat different problems, but as both are related to aspects of cache & performance, just wanted to link them for future reference. In this sense, after some investigation, I have an hypothesis of what could be causing the problem described in this issue: in summary, _it may be the case that we need to increase the parallelism of RocksDB compaction_.

_Symptoms:_
- The problem only occurs with unthrottled clients.
- The problem manifests with a sudden increase of GET request latency on RocksDB. This, in turn, makes reads from `StorageWriter` to be slower, and therefore slow down the throughput of Pravega to Tier 2.
- In our bare metal cluster, we observed that in this situation, the cache drive presents a very high utilization, which is induced mainly by reads (read amplification).

_Hypothesis:_

In Pravega, RocksDB needs to perform compaction on deleted keys, and we have a lot of deletes given that the `CacheManager` is continuously evicting data from the cache (see https://github.com/facebook/rocksdb/wiki/RocksDB-Basics#multi-threaded-compactions):
> Compactions are needed to remove multiple copies of the same key that may occur if an application overwrites an existing key. Compactions also process deletions of keys. Compactions may occur in multiple threads if configured appropriately.

However, we do not configure by default parallel threads for compaction. As can be read in the same page, it looks as a key point for the long-term performance of the database, specially during heavy, sustained workloads:
> The overall write throughput of an LSM database directly depends on the speed at which compactions can occur, especially when the data is stored in fast storage like SSD or RAM. RocksDB may be configured to issue concurrent compaction requests from multiple threads. It is observed that sustained write rates may increase by as much as a factor of 10 with multi-threaded compaction when the database is on SSDs, as compared to single-threaded compactions.

By default, there is only 1 thread doing compaction (and 1 doing subcompactions from L0 -> L1), which in turn belongs to the LOW priority thread pool. On the other hand, the thread doing flush belongs to the HIGH priority thread pool. Therefore, having a single, low-priority thread to execute compactions may be not enough. This may be specially true considering unthrottled clients, as they execute a continuos, high rate of writes to RocksDB (and reads from `StorageWriter`) that can greatly reduce the execution time for the compaction thread to carry out its job. As a consequence of the insufficient compaction activity, we may be suffering a very high read amplification that leads the drive to present high utilization and GETs to be very slow. The rest of symptoms are described in this issue.

I have created a branch with new RocksDB settings to increase its internal compaction parallelism and we will validate this hypothesis as soon as possible.
